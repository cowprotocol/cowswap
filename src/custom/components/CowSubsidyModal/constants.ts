// used in modal and popover
export const SUBSIDY_INFO_MESSAGE = '(v)COW token holders are eligible for a fee discount'
export const COW_SUBSIDY_DATA: [string, number][] = [
  ['0', 0],

  // 1,000
  ['1000000000000000000000', 5],

  // 10,000
  ['10000000000000000000000', 10],

  // 100,000
  ['100000000000000000000000', 20],

  // 1,000,000
  ['1000000000000000000000000', 40],
]
