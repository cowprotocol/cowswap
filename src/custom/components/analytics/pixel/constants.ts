export enum PIXEL_EVENTS {
  INIT = 'init',
  CONNECT_WALLET = 'connect_wallet',
  POST_TRADE = 'post_trade',
}
