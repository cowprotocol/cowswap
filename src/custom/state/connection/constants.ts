import { ConnectionType } from 'connection'

export const BACKFILLABLE_WALLETS = [ConnectionType.INJECTED]
