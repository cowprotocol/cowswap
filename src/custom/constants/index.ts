import { Token, Fraction, Percent } from '@uniswap/sdk-core'

import { GPv2Settlement, GPv2VaultRelayer } from '@gnosis.pm/gp-v2-contracts/networks.json'
import { WalletInfo, SUPPORTED_WALLETS as SUPPORTED_WALLETS_UNISWAP } from 'constants/wallet'

import { SupportedChainId as ChainId } from 'constants/chains'
import { getAppDataHash } from './appDataHash'

export const INITIAL_ALLOWED_SLIPPAGE_PERCENT = new Percent('5', '1000') // 0.5%
export const RADIX_DECIMAL = 10
export const RADIX_HEX = 16

export const DEFAULT_DECIMALS = 18
export const DEFAULT_PRECISION = 6
export const DEFAULT_SMALL_LIMIT = '0.000001'
export const AMOUNT_PRECISION = 4
export const LONG_PRECISION = 10
export const FULL_PRICE_PRECISION = 20
export const FIAT_PRECISION = 2
export const PERCENTAGE_PRECISION = 2

export const SHORT_LOAD_THRESHOLD = 500
export const LONG_LOAD_THRESHOLD = 2000

export const APP_DATA_HASH = getAppDataHash()
export const PRODUCTION_URL = 'cowswap.exchange'
export const BARN_URL = `barn.${PRODUCTION_URL}`

// Allow WALLET_LINK to be activated on mobile
// since COINBASE_LINK is limited to use only 1 deeplink on mobile
SUPPORTED_WALLETS_UNISWAP.WALLET_LINK = {
  ...SUPPORTED_WALLETS_UNISWAP.WALLET_LINK,
  mobile: true,
}
const DISABLED_WALLETS = /^(?:Portis|COINBASE_LINK)$/i

// Re-export only the supported wallets
export const SUPPORTED_WALLETS = Object.keys(SUPPORTED_WALLETS_UNISWAP).reduce((acc, key) => {
  if (!DISABLED_WALLETS.test(key)) {
    acc[key] = SUPPORTED_WALLETS_UNISWAP[key]
  }
  return acc
}, {} as { [key: string]: WalletInfo })

// Smart contract wallets are filtered out by default, no need to add them to this list
export const UNSUPPORTED_WC_WALLETS = new Set(['DeFi Wallet', 'WallETH'])

export const GP_SETTLEMENT_CONTRACT_ADDRESS: Partial<Record<number, string>> = {
  [ChainId.MAINNET]: GPv2Settlement[ChainId.MAINNET].address,
  [ChainId.RINKEBY]: GPv2Settlement[ChainId.RINKEBY].address,
  [ChainId.XDAI]: GPv2Settlement[ChainId.XDAI].address,
}

export const GP_VAULT_RELAYER: Partial<Record<number, string>> = {
  [ChainId.MAINNET]: GPv2VaultRelayer[ChainId.MAINNET].address,
  [ChainId.RINKEBY]: GPv2VaultRelayer[ChainId.RINKEBY].address,
  [ChainId.XDAI]: GPv2VaultRelayer[ChainId.XDAI].address,
}

export const V_COW_CONTRACT_ADDRESS: Record<number, string> = {
  [ChainId.MAINNET]: '0xd057b63f5e69cf1b929b356b579cba08d7688048',
  [ChainId.XDAI]: '0xc20C9C13E853fc64d054b73fF21d3636B2d97eaB',
  [ChainId.RINKEBY]: '0x9386177e95A853070076Df2403b9D547D653126D',
}

export const COW_CONTRACT_ADDRESS: Record<number, string> = {
  [ChainId.MAINNET]: '0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB',
  [ChainId.XDAI]: '0x177127622c4A00F3d409B75571e12cB3c8973d3c',
  [ChainId.RINKEBY]: '0xbdf1e19f8c78A77fb741b44EbA5e4c0C8DBAeF91',
}

// See https://github.com/gnosis/gp-v2-contracts/commit/821b5a8da213297b0f7f1d8b17c893c5627020af#diff-12bbbe13cd5cf42d639e34a39d8795021ba40d3ee1e1a8282df652eb161a11d6R13
export const NATIVE_CURRENCY_BUY_ADDRESS = '0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE'
export const NATIVE_CURRENCY_BUY_TOKEN: { [chainId in ChainId | number]: Token } = {
  [ChainId.MAINNET]: new Token(ChainId.MAINNET, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'ETH', 'Ether'),
  [ChainId.RINKEBY]: new Token(ChainId.RINKEBY, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'ETH', 'Ether'),
  // [ChainId.ROPSTEN]: new Token(ChainId.ROPSTEN, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'ETH', 'Ether'),
  // [ChainId.GOERLI]: new Token(ChainId.GOERLI, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'ETH', 'Ether'),
  // [ChainId.KOVAN]: new Token(ChainId.KOVAN, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'ETH', 'Ether'),
  [ChainId.XDAI]: new Token(ChainId.XDAI, NATIVE_CURRENCY_BUY_ADDRESS, 18, 'xDAI', 'xDAI'),
}

export const ORDER_ID_SHORT_LENGTH = 8
export const INPUT_OUTPUT_EXPLANATION = 'Only executed swaps incur fees.'
export const DEFAULT_ORDER_DELAY = 20000 // 20s
export const PENDING_ORDERS_BUFFER = 60 * 1000 // 60s
export const CANCELLED_ORDERS_PENDING_TIME = 5 * 60 * 1000 // 5min
export const PRICE_API_TIMEOUT_MS = 10000 // 10s
export const GP_ORDER_UPDATE_INTERVAL = 30 * 1000 // 30s
export const MINIMUM_ORDER_VALID_TO_TIME_SECONDS = 120

export const WETH_LOGO_URI =
  'https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/logo.png'
export const XDAI_LOGO_URI =
  'https://raw.githubusercontent.com/1Hive/default-token-list/master/src/assets/xdai/0xe91d153e0b41518a2ce8dd3d7944fa863463a97d/logo.png'

export const DOCS_LINK = 'https://docs.cow.fi'
export const CONTRACTS_CODE_LINK = 'https://github.com/gnosis/gp-v2-contracts'
export const CODE_LINK = 'https://github.com/cowprotocol/cowswap'
export const DISCORD_LINK = 'https://discord.com/invite/cowprotocol'
export const DUNE_DASHBOARD_LINK = 'https://duneanalytics.com/gnosis.protocol/Gnosis-Protocol-V2'
export const TWITTER_LINK = 'https://twitter.com/mevprotection'
export const GPAUDIT_LINK = 'https://github.com/gnosis/gp-v2-contracts/blob/main/audits/GnosisProtocolV2May2021.pdf'
export const FLASHBOYS_LINK = 'https://arxiv.org/abs/1904.05234'
export const COWWIKI_LINK = 'https://en.wikipedia.org/wiki/Coincidence_of_wants'
export const GNOSIS_FORUM_ROADTODECENT_LINK = 'https://forum.gnosis.io/t/gpv2-road-to-decentralization/1245'

// MEV Metrics from https://explore.flashbots.net/
export const MEV_TOTAL = '606 Million'
export const FLASHBOTS_LINK = 'https://explore.flashbots.net/'

// 30 minutes
export const GAS_PRICE_UPDATE_THRESHOLD = 30 * 60 * 1000
export const GAS_FEE_ENDPOINTS = {
  [ChainId.MAINNET]: 'https://safe-relay.gnosis.io/api/v1/gas-station/',
  // No ropsten = main
  // [ChainId.ROPSTEN]: 'https://safe-relay.gnosis.io/api/v1/gas-station/',
  [ChainId.RINKEBY]: 'https://safe-relay.rinkeby.gnosis.io/api/v1/gas-station/',
  // [ChainId.GOERLI]: 'https://safe-relay.goerli.gnosis.io/api/v1/gas-station/',
  // no kovan = main
  // [ChainId.KOVAN]: 'https://safe-relay.kovan.gnosis.io/api/v1/gas-station/',
  [ChainId.XDAI]: 'https://blockscout.com/xdai/mainnet/api/v1/gas-price-oracle',
}

export const UNSUPPORTED_TOKENS_FAQ_URL = '/faq/trading#what-token-pairs-does-cowswap-allow-to-trade'

// fee threshold - should be greater than percentage, show warning
export const FEE_SIZE_THRESHOLD = new Fraction(10, 100) // 30%

// default value provided as userAddress to Paraswap API if the user wallet is not connected
export const SOLVER_ADDRESS = '0xa6ddbd0de6b310819b49f680f65871bee85f517e'

export const MAXIMUM_ORDERS_TO_DISPLAY = 10
export const AMOUNT_OF_ORDERS_TO_FETCH = 100

// last wallet provider key used in local storage
export const STORAGE_KEY_LAST_PROVIDER = 'lastProvider'
export const WAITING_TIME_RECONNECT_LAST_PROVIDER = 15000 // 15s

// Default price strategy to use for getting app prices
// COWSWAP = new quote endpoint
// LEGACY = price racing logic (checking 0x, gp, paraswap, etc)
export const DEFAULT_GP_PRICE_STRATEGY = 'COWSWAP'

// Start date of COW vesting for locked GNO
export const LOCKED_GNO_VESTING_START_DATE = new Date('02-11-2022 13:05:15 GMT')

// These values match the vesting contract configuration (see contract's `cliffTime` and `duration` fields).
// They are fixed and will never change.
export const LOCKED_GNO_VESTING_START_TIME = 1644584715000
export const LOCKED_GNO_VESTING_DURATION = 126144000000 // 4 years
