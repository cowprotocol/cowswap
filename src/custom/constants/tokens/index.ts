import { ChainId } from '@uniswap/sdk'
import { WETH9, Token } from '@uniswap/sdk-core'
import { DAI_RINKEBY, USDC_RINKEBY, USDT_RINKEBY, WBTC_RINKEBY } from 'utils/rinkeby/constants'
import { DAI, USDC_MAINNET, USDT, WBTC } from '@src/constants/tokens'
import { USDC_XDAI, /*USDT_XDAI,*/ WBTC_XDAI, WETH_XDAI, WXDAI } from 'utils/xdai/constants'
import { SupportedChainId } from 'constants/chains'
import { V_COW_CONTRACT_ADDRESS, COW_CONTRACT_ADDRESS } from 'constants/index'

import wxDaiLogo from 'assets/cow-swap/wxdai.png'
// TODO: these are the same? why?
import vCowLogo from 'assets/cow-swap/vCOW.png'
import cowLogo from 'assets/cow-swap/cow.svg'
import gnoLogo from 'assets/cow-swap/gno.png'
import usdcLogo from 'assets/cow-swap/usdc.png'

export * from './tokensMod'

function getTrustImage(mainnetAddress: string): string {
  return `https://raw.githubusercontent.com/trustwallet/assets/master/blockchains/ethereum/assets/${mainnetAddress}/logo.png`
}

const WETH_ADDRESS_MAINNET = WETH9[ChainId.MAINNET].address

/**
 * vCow token
 */
const V_COW_TOKEN_MAINNET = new Token(
  SupportedChainId.MAINNET,
  V_COW_CONTRACT_ADDRESS[SupportedChainId.MAINNET] || '',
  18,
  'vCOW',
  'CoW Protocol Virtual Token'
)

const V_COW_TOKEN_XDAI = new Token(
  SupportedChainId.XDAI,
  V_COW_CONTRACT_ADDRESS[SupportedChainId.XDAI] || '',
  18,
  'vCOW',
  'CoW Protocol Virtual Token'
)

const V_COW_TOKEN_RINKEBY = new Token(
  SupportedChainId.RINKEBY,
  V_COW_CONTRACT_ADDRESS[SupportedChainId.RINKEBY] || '',
  18,
  'vCOW',
  'CoW Protocol Virtual Token'
)

export const V_COW: Record<number, Token> = {
  [SupportedChainId.MAINNET]: V_COW_TOKEN_MAINNET,
  [SupportedChainId.XDAI]: V_COW_TOKEN_XDAI,
  [SupportedChainId.RINKEBY]: V_COW_TOKEN_RINKEBY,
}

/**
 * Cow token
 */
const COW_TOKEN_MAINNET = new Token(
  SupportedChainId.MAINNET,
  COW_CONTRACT_ADDRESS[SupportedChainId.MAINNET] || '',
  18,
  'COW',
  'CoW Protocol Token'
)

const COW_TOKEN_XDAI = new Token(
  SupportedChainId.XDAI,
  COW_CONTRACT_ADDRESS[SupportedChainId.XDAI] || '',
  18,
  'COW',
  'CoW Protocol Token'
)

const COW_TOKEN_RINKEBY = new Token(
  SupportedChainId.RINKEBY,
  COW_CONTRACT_ADDRESS[SupportedChainId.RINKEBY] || '',
  18,
  'COW',
  'CoW Protocol Token'
)

export const COW: Record<number, Token> = {
  [SupportedChainId.MAINNET]: COW_TOKEN_MAINNET,
  [SupportedChainId.XDAI]: COW_TOKEN_XDAI,
  [SupportedChainId.RINKEBY]: COW_TOKEN_RINKEBY,
}

/**
 * GNO token
 */
const GNO_MAINNET = new Token(
  SupportedChainId.MAINNET,
  '0x6810e776880c02933d47db1b9fc05908e5386b96',
  18,
  'GNO',
  'Gnosis'
)
const GNO_XDAI = new Token(SupportedChainId.XDAI, '0x9c58bacc331c9aa871afd802db6379a98e80cedb', 18, 'GNO', 'Gnosis')
const GNO_RINKEBY = new Token(
  SupportedChainId.RINKEBY,
  '0xd0dab4e640d95e9e8a47545598c33e31bdb53c7c',
  18,
  'GNO',
  'Gnosis'
)

export const GNO: Record<SupportedChainId, Token> = {
  [SupportedChainId.MAINNET]: GNO_MAINNET,
  [SupportedChainId.XDAI]: GNO_XDAI,
  [SupportedChainId.RINKEBY]: GNO_RINKEBY,
}

export const ADDRESS_IMAGE_OVERRIDE = {
  // Rinkeby
  [DAI_RINKEBY.address]: getTrustImage(DAI.address),
  [USDC_RINKEBY.address]: getTrustImage(USDC_MAINNET.address),
  [USDT_RINKEBY.address]: getTrustImage(USDT.address),
  [WBTC_RINKEBY.address]: getTrustImage(WBTC.address),
  [WETH9[ChainId.RINKEBY].address]: getTrustImage(WETH_ADDRESS_MAINNET),
  [V_COW_TOKEN_RINKEBY.address]: vCowLogo,
  [COW_TOKEN_RINKEBY.address]: cowLogo,
  [GNO_RINKEBY.address]: gnoLogo,
  [USDC_RINKEBY.address]: usdcLogo,
  // xDai
  [USDC_XDAI.address]: getTrustImage(USDC_MAINNET.address),
  // [USDT_XDAI.address]: getTrustImage(USDT.address),
  [WBTC_XDAI.address]: getTrustImage(WBTC.address),
  [WXDAI.address]: wxDaiLogo,
  [WETH_XDAI.address]: getTrustImage(WETH_ADDRESS_MAINNET),
  [V_COW_TOKEN_XDAI.address]: vCowLogo,
  [COW_TOKEN_XDAI.address]: cowLogo,
  [GNO_XDAI.address]: gnoLogo,
  [USDC_XDAI.address]: usdcLogo,
  // Mainnet
  [V_COW_TOKEN_MAINNET.address]: vCowLogo,
  [COW_TOKEN_MAINNET.address]: cowLogo,
}

/**
 * Addresses related to COW vesting for Locked GNO
 * These are used in src/custom/pages/Profile/LockedGnoVesting hooks and index files
 */
export const MERKLE_DROP_CONTRACT_ADDRESSES: Record<number, string> = {
  [SupportedChainId.MAINNET]: '0x64646f112FfD6F1B7533359CFaAF7998F23C8c40',
  [SupportedChainId.RINKEBY]: '0x5444c4AFb2ec7f7367C10F7732b8558650c5899F',
  [SupportedChainId.XDAI]: '0x3d610e917130f9D036e85A030596807f57e11093',
}

export const TOKEN_DISTRO_CONTRACT_ADDRESSES: Record<number, string> = {
  [SupportedChainId.MAINNET]: '0x68FFAaC7A431f276fe73604C127Bd78E49070c92',
  [SupportedChainId.RINKEBY]: '0xeBA8CE5b23c054f1511F8fF5114d848329B8258d',
  [SupportedChainId.XDAI]: '0x3d610e917130f9D036e85A030596807f57e11093',
}
