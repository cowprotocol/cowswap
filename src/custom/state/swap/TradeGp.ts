import { CanonicalMarketParams, getCanonicalMarket } from 'utils/misc'
import { CurrencyAmount, Currency, TradeType, Price, Percent } from '@uniswap/sdk-core'
import { Trade } from '@uniswap/v2-sdk'
import { FeeInformation, PriceInformation } from '@cow/types'
import { ONE_FRACTION } from 'constants/misc'

export type FeeForTrade = { feeAsCurrency: CurrencyAmount<Currency> } & Pick<FeeInformation, 'amount'>

export type TradeWithFee = Omit<Trade<Currency, Currency, TradeType>, 'nextMidPrice' | 'exactIn' | 'exactOut'> & {
  inputAmountWithFee: CurrencyAmount<Currency>
  outputAmountWithoutFee?: CurrencyAmount<Currency>
  fee: FeeForTrade
}

type TradeExecutionPrice = CanonicalMarketParams<CurrencyAmount<Currency> | undefined> & { price?: PriceInformation }

export function _constructTradePrice({
  sellToken,
  buyToken,
  kind,
  price,
}: TradeExecutionPrice): Price<Currency, Currency> | undefined {
  if (!sellToken || !buyToken || !price?.amount) return

  let executionPrice: Price<Currency, Currency> | undefined
  // get canonical market tokens
  // to accurately create our Price<Currency, Currency>
  const { baseToken, quoteToken } = getCanonicalMarket({
    sellToken,
    buyToken,
    kind,
  })

  if (baseToken && quoteToken && price) {
    executionPrice = new Price<Currency, Currency>(
      // baseToken.currency,
      // quoteToken.currency,
      // baseToken.currency.quotient,
      // price.amount
      // TODO: CHECK THIS IS THE SAME AS THE ABOVE ON THE OLDER SDK
      { baseAmount: baseToken, quoteAmount: CurrencyAmount.fromRawAmount(quoteToken.currency, price.amount) }
    )
  }
  return executionPrice
}

export function _minimumAmountOut(pct: Percent, trade: TradeGp) {
  if (trade.tradeType === TradeType.EXACT_OUTPUT) {
    return trade.outputAmount
  }

  return trade.outputAmount.multiply(ONE_FRACTION.subtract(pct))
}

export function _maximumAmountIn(pct: Percent, trade: TradeGp) {
  if (trade.tradeType === TradeType.EXACT_INPUT) {
    return trade.inputAmount
  }

  return trade.inputAmountWithFee.multiply(ONE_FRACTION.add(pct))
}

interface TradeGpConstructor {
  inputAmount: CurrencyAmount<Currency>
  inputAmountWithFee: CurrencyAmount<Currency>
  inputAmountWithoutFee: CurrencyAmount<Currency>
  outputAmount: CurrencyAmount<Currency>
  outputAmountWithoutFee: CurrencyAmount<Currency>
  fee: FeeForTrade
  executionPrice: Price<Currency, Currency>
  tradeType: TradeType
  quoteId?: number
}

/**
 * Represents a trade executed against a list of pairs.
 * Does not account for slippage, i.e. trades that front run this trade and move the price.
 */
export default class TradeGp {
  /**
   * The type of the trade, either exact in or exact out.
   */
  readonly tradeType: TradeType
  /**
   * The input amount for the trade assuming no slippage.
   */
  readonly inputAmount: CurrencyAmount<Currency>
  readonly inputAmountWithFee: CurrencyAmount<Currency>
  readonly inputAmountWithoutFee: CurrencyAmount<Currency>
  /**
   * The output amount for the trade assuming no slippage.
   */
  readonly outputAmount: CurrencyAmount<Currency>
  readonly outputAmountWithoutFee?: CurrencyAmount<Currency>
  /**
   * Trade fee
   */
  readonly fee: FeeForTrade
  /**
   * The price expressed in terms of output amount/input amount.
   */
  readonly executionPrice: Price<Currency, Currency>

  /**
   * The id returned by CoW Swap's quote backend, if any
   *
   * Note that it won't be set for fast quotes, nor for quotes from other sources (paraswap, 0x, etc)
   */
  readonly quoteId?: number

  public constructor({
    inputAmount,
    inputAmountWithFee,
    inputAmountWithoutFee,
    outputAmount,
    outputAmountWithoutFee,
    fee,
    executionPrice,
    tradeType,
    quoteId,
  }: TradeGpConstructor) {
    this.tradeType = tradeType
    this.inputAmount = inputAmount
    this.inputAmountWithFee = inputAmountWithFee
    this.inputAmountWithoutFee = inputAmountWithoutFee
    this.outputAmountWithoutFee = outputAmountWithoutFee
    this.outputAmount = outputAmount
    this.fee = fee
    this.executionPrice = executionPrice
    this.quoteId = quoteId
  }
  /**
   * Get the minimum amount that must be received from this trade for the given slippage tolerance
   * @param slippageTolerance tolerance of unfavorable slippage from the execution price of this trade
   */
  public minimumAmountOut(slippageTolerance: Percent): CurrencyAmount<Currency> {
    return _minimumAmountOut(slippageTolerance, this)
  }
  /**
   * Get the maximum amount in that can be spent via this trade for the given slippage tolerance
   * @param slippageTolerance tolerance of unfavorable slippage from the execution price of this trade
   */
  public maximumAmountIn(slippageTolerance: Percent): CurrencyAmount<Currency> {
    return _maximumAmountIn(slippageTolerance, this)
  }
}
