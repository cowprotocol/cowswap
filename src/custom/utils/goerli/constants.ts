import { Token } from '@uniswap/sdk-core'
import { SupportedChainId as ChainId } from 'constants/chains'

// Todo: address should be updated
export const DAI_GOERLI = new Token(ChainId.GOERLI, '0x5c221e77624690fff6dd741493d735a17716c26b', 18, 'DAI', 'DAI')

export const USDT_GOERLI = new Token(
  ChainId.GOERLI,
  '0xe583769738b6dd4e7caf8451050d1948be717679',
  6,
  'USDT',
  'Tether USD'
)
export const USDC_GOERLI = new Token(
  ChainId.GOERLI,
  '0x5ddba296458986e462ba4aeb4dddc00fadc9c43b',
  6,
  'USDC',
  'USD Coin'
)
export const WBTC_GOERLI = new Token(
  ChainId.GOERLI,
  '0xca063a2ab07491ee991dcecb456d1265f842b568',
  8,
  'WBTC',
  'Wrapped BTC'
)
