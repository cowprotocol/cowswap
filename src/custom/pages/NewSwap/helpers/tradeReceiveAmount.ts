import { TradeType } from '@uniswap/sdk-core'
import { formatSmartAmount } from 'utils/format'
import TradeGp from 'state/swap/TradeGp'

export interface ReceiveAmountInfo {
  type: 'from' | 'to'
  amountBeforeFees: string
  amountAfterFees: string
  feeAmount: string
}

export function getInputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'from',
    amountBeforeFees:
      trade.tradeType === TradeType.EXACT_INPUT && trade.inputAmountWithFee.lessThan(trade.fee.amount)
        ? '0'
        : formatSmartAmount(trade.inputAmountWithoutFee) || '0',
    amountAfterFees: formatSmartAmount(trade.inputAmountWithFee) || '0',
    feeAmount: formatSmartAmount(trade.fee.feeAsCurrency) || '0',
  }
}

export function getOutputReceiveAmountInfo(trade: TradeGp): ReceiveAmountInfo {
  return {
    type: 'to',
    amountBeforeFees: formatSmartAmount(trade.outputAmountWithoutFee) || '0',
    amountAfterFees: formatSmartAmount(trade.outputAmount) || '0',
    feeAmount: formatSmartAmount(trade.outputAmountWithoutFee?.subtract(trade.outputAmount)) || '0',
  }
}
