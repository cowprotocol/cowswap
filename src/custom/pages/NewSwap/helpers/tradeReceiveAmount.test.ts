import { getInputReceiveAmountInfo, getOutputReceiveAmountInfo } from './tradeReceiveAmount'
import TradeGp from 'state/swap/TradeGp'
import { COW, GNO } from 'constants/tokens'
import { SupportedChainId } from 'constants/chains'
import { CurrencyAmount, Price, TradeType } from '@uniswap/sdk-core'

const currency = COW[SupportedChainId.MAINNET]
const currencyOut = GNO[SupportedChainId.MAINNET]
const amount = 250
const output = 250

describe('Helpers to build ReceiveAmountInfo', () => {
  describe('getInputReceiveAmountInfo()', () => {
    describe('When tradeType is EXACT_INPUT and inputAmountWithFee less than trade fee amount', () => {
      it('Then amountBeforeFees should be zero', () => {
        const value = getInputReceiveAmountInfo(
          new TradeGp({
            inputAmount: CurrencyAmount.fromRawAmount(currency, amount * 10 ** 18),
            inputAmountWithFee: CurrencyAmount.fromRawAmount(currency, 0),
            inputAmountWithoutFee: CurrencyAmount.fromRawAmount(currency, amount * 10 ** 18),
            outputAmount: CurrencyAmount.fromRawAmount(currency, output * 10 ** 18),
            outputAmountWithoutFee: CurrencyAmount.fromRawAmount(currency, (output - 3) * 10 ** 18),
            fee: { feeAsCurrency: CurrencyAmount.fromRawAmount(currency, 3 * 10 ** 18), amount: '5000' },
            executionPrice: new Price(currency, currencyOut, 1, 4),
            tradeType: TradeType.EXACT_INPUT,
            quoteId: 10000,
          })
        )

        expect(value).toMatchSnapshot()
        expect(value.amountBeforeFees).toBe('0')
      })
    })
  })

  describe('getOutputReceiveAmountInfo()', () => {
    it('Should match a snapshot', () => {
      const value = getOutputReceiveAmountInfo(
        new TradeGp({
          inputAmount: CurrencyAmount.fromRawAmount(currency, amount * 10 ** 18),
          inputAmountWithFee: CurrencyAmount.fromRawAmount(currency, 0),
          inputAmountWithoutFee: CurrencyAmount.fromRawAmount(currency, amount * 10 ** 18),
          outputAmount: CurrencyAmount.fromRawAmount(currency, output * 10 ** 18),
          outputAmountWithoutFee: CurrencyAmount.fromRawAmount(currency, (output - 3) * 10 ** 18),
          fee: { feeAsCurrency: CurrencyAmount.fromRawAmount(currency, 3 * 10 ** 18), amount: '5000' },
          executionPrice: new Price(currency, currencyOut, 1, 4),
          tradeType: TradeType.EXACT_INPUT,
          quoteId: 10000,
        })
      )

      expect(value).toMatchSnapshot()
    })
  })
})
