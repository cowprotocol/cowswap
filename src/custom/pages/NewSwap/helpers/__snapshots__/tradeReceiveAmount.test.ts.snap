// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers to build ReceiveAmountInfo getInputReceiveAmountInfo() When tradeType is EXACT_INPUT and inputAmountWithFee less than trade fee amount Then amountBeforeFees should be zero 1`] = `
Object {
  "amountAfterFees": "0",
  "amountBeforeFees": "0",
  "feeAmount": "3",
  "type": "from",
}
`;

exports[`Helpers to build ReceiveAmountInfo getOutputReceiveAmountInfo() Should match a snapshot 1`] = `
Object {
  "amountAfterFees": "250",
  "amountBeforeFees": "247",
  "feeAmount": "-3",
  "type": "to",
}
`;
